tests/usb_interfaces.test.cpp

```diff
@@ -0,0 +1,737 @@
+
+    std::array<byte const, 1> iface_two_arr({ m_iface_two.num });
+    std::array<byte const, 1> iface_two_str_arr({ m_iface_two.str_idx });
+
+    auto span_arr = make_scatter_bytes(iad_buf,
+
+                                       // iface one
+                                       iface_header,
+                                       std::span(iface_one_arr),
+                                       static_desc_vars,
+                                       std::span(iface_one_str_arr),
+
+                                       // iface two
+                                       iface_header,
+                                       std::span(iface_two_arr),
+                                       static_desc_vars,
+                                       std::span(iface_two_str_arr));
+
+    p_callback(span_arr);
+    m_wrote_descriptors = true;
+    return { 2, 3 };
+  }
+
+  [[nodiscard]] bool driver_write_string_descriptor(
+    endpoint_writer const& p_callback,
+    u8 p_index) override
+  {
+    if (!m_wrote_descriptors) {
+      // std::println("You must call write_descriptors first");
+      safe_throw(hal::operation_not_permitted(this));
+    }
+    std::array<byte, 2> header{ 0, str_desc_type };
+    if (p_index == m_iface_one.str_idx) {
+      header[0] = m_name_one.length() + 2;
+      auto arr = make_scatter_bytes(
+        header,
+        std::span(reinterpret_cast<byte const*>(m_name_one.data()),
+                  m_name_one.length()));
+      p_callback(arr);
+      return true;
+    } else if (p_index == m_iface_two.str_idx) {
+      header[0] = m_iface_two.str_idx + 2;
+      auto arr = make_scatter_bytes(
+        header,
+        std::span(reinterpret_cast<byte const*>(m_name_two.data()),
+                  m_name_two.length()));
+
+      p_callback(arr);
+      return true;
+    }
+
+    return false;
+  }
+
+  [[nodiscard]] u16 driver_get_status(req_bitmap p_bmRequestType,
+                                      u16 p_wIndex) override
+  {
+    if (p_bmRequestType.get_recipient() != req_bitmap::recipient::interface) {
+      // std::println("Unsupported recipient");
+      safe_throw(hal::operation_not_supported(this));
+    }
+
+    auto iface_idx = p_wIndex & 0xFF;
+
+    if (iface_idx == m_iface_one.num) {
+      return m_iface_one.num;
+    } else if (iface_idx == m_iface_two.num) {
+      return m_iface_two.num;
+    }
+
+    // std::println("Invalid interface index");
+    safe_throw(hal::operation_not_supported(this));
+  }
+
+  void driver_manage_features(req_bitmap p_bmRequestType,
+                              bool p_set,
+                              u16 p_selector,
+                              u16 p_wIndex) override
+  {
+    std::ignore = p_selector;
+    if (p_bmRequestType.get_recipient() != req_bitmap::recipient::interface) {
+      // std::println("Unsupported recipient");
+      safe_throw(hal::operation_not_supported(this));
+    }
+
+    auto iface_idx = p_wIndex & 0xFF;
```

Comment:
Ooooo, this is endianness specific isn't it? This might be a reason to add this as an API in the request struct. But because this is so critical to USB, I'd say make it a helper function within that struct as well.